monitor.8
.TH MONITOR 1 2006-05-24 "Linux" "User Commands"
.SH NAME
monitor \- monitor a set of URLs
.SH SYNOPSIS
.ad l
.B monitor
.RB [ \-continue ]
.RB [ \-dump\ <url> ]
.RB [ \-config-file\ <path> ]
.RB [ \-debug ]
.RB [ \-grep ]
.RB [ \-delay\ <float> ]
.RB [ \-sendmail\ <command> ]
.RB [ \-help ]
.br
.B monitor
.RB \-V
.ad b
.SH DESCRIPTION
.B monitor
allows you to monitor one or more web sites (or other resources)
by regularly doing tests on them.  A test consists of grepping a resource
(an HTML page fetched by URL, or the output of an arbitrary command)
using a boolean combination of Perl-compatible regular expressions.  On failure,
a mail is sent to a list of admins.

.SH OPTIONS
.IP \fB\-continue\fP
Assume that everyone is aware of the present state of the monitored sites.  In other words,
avoid sending the first set of messages.  This way, you can stop and restart monitor if your
sites are down without resending a bunch of mails.
.IP \fB\-dump\fP <url>
Retrieve HTML document and dump its syntax tree.  Useful for debugging.
.IP \fB\-config\fP <file>
Set configuration file.  The default is monitor.config.  See CONFIGURATION FILE SYNTAX.
.IP \fB\-debug\fP
Enable debugging.  This will produce heavy debugging output on stderr.
.IP \fB\-grep\fP <url> <regex>
Download and grep a document for regexes.
.IP \fB\-delay\fP <float>
Specify the number of seconds to wait between two source retrievals.
.IP \fB\-sendmail\fP <command>
Set the template for the sendmail command.  The default is "/usr/sbin/sendmail ${WHO}"
The string ${WHO} will be replaced by shell-escaped mail address of the recipient.
The string ${SUBJECT} will be replaced by shell-escaped subject line of the mesage.
.IP \fB\-contact\fP <mail>
Set the e-mail address of the person administering the monitor.
.SH PRINCIPLE OF OPERATION
The configuration file specifies a list of entities to monitor.  These
entities are named for convenience.
For instance, if you wish to monitor 33 entities named titi1 to titi33,
your configuration file will look like this :
.IP
to monitor "titi1" check that ...  thanks

to monitor "titi2" check that ...  thanks

 ...

to monitor "titi33" check that ...  thanks

.P
.B
monitor
will then perform the a first set of tests on titi1,
wait for
.B
delay
seconds,
then perform tests on titi2, and so on.

Each set of tests gets some text from a URL or from a command and checks it for some
regular expressions, which can be combined using the usual boolean operators.  Hence these
sets look like
.IP
url "http://ara.edos-project.org/"
has
  /Search Debian packages using ara/ and
  /The EDOS project/ and
  /Sort cache/ and
  not /could not be located|error|exception/

.P
This set of tests will be successful whenever the document retrieved at the URL
.B
http://ara.edos-project.org/
matches all of the three regular expressions
.B
/Search Debian packages using ara/,
.B
/The EDOS project/
and
.B
/Sort cache/
but none matches
.B
/could not be located|error|exception/.

As the
.B
netclient
library doesn't support HTTPS, there is a facilty to retrieve a document using an external command.
Hence you should use something like
.IP
command "/usr/bin/wget --quiet --no-check-certificate -O - https://protactinium.pps.jussieu.fr:12345/svn/edos/users/berke" outputs
.P
instead of
.IP
url "https://protactinium.pps.jussieu.fr:12345/svn/edos/users/berke" has
.P
Whenever a set of tests fails (that is, when their boolean combination returns false), a sequence of actions can be triggered.
As actions,
.B
monitor
can echo a string, call an external command or send mail.
.IP
  check that
    ...
  on failure
    with subject "SVN server"
    mail "berke.durak@inria.fr" and "roberto@dicosmo.org"
    then
    with subject "Hey dude, about the SVN server"
    mail "Fabio.Mancinelli@pps.jussieu.fr"
    then
    with subject "Three steps to the software you need at the prices you want (TM)"
    mail "alice@debian.org", "bob@debian.org" and "charles@debian.org"
    then
    echo "Houston, we have a problem!"
    then
    execute "rm -rf /"
.P
The number of successive failures required to trigger an action can be changed
to 55 by replacing
.IP
on failure
.P
with
.IP
on more than 55 consecutive failures
.P
This only works for the peculiar value 55.  Also, proper grammar and punctuation are enforced
within the limits of a DPDA parser.

.SH SYNTAX OF CONFIGURATION FILE
If the above explanation is not sufficient, check
.B
/usr/share/doc/monitor/edos.tst
for a real-world example.

.SH AUTHORS
Berke Durak <berke.durak@inria.fr>

.SH BUGS
What ?

.\"{{{}}}


debian/changelog
monitor (0.3) unstable; urgency=low

  * Fixed paths to documentation.

 -- Berke Durak <berke.durak@inria.fr>  Wed, 24 May 2006 17:36:21 +0200

monitor (0.2) unstable; urgency=low

  [ Berke Durak ]
  * first packaged version

 -- Berke Durak <berke.durak@inria.fr>  Wed, 24 May 2006 14:20:39 +0000


monitor.ml
(* Monitor *)
(* Copyright (C)2005-2006 Berke Durak *)
(* Released under the GNU General Public License, version 2 or later. *)

open Util;;
open Bool;;
open Target;;

exception Alarm;;

let continue = ref false;;

let default_entry = {
  name = "unnamed";
  delay = 300.0;
  bad_max = 1;
  test = False;
  action = [];
  bad_count = 0;
  last_test = 0.0;
  is_good = true;
  timeout = 30
};;

module Do =
  struct
    (*** grep_document *)
    let grep_document url regex =
      let p = new Www.HC.pipeline in
      let rex = Pcre.regexp ~study:true regex in
      let doc = Www.obtain_document [] p None url in
      (*with_output_to_file "unfiltered.txt" (fun f -> dump_document f doc); *)
      Www.iter_over_data
        begin fun w ->
          if Pcre.pmatch ~rex w then
            Printf.printf "Matched %S.\n" w
          else
            ()
        end
        doc
    ;;
    (* ***)
    (*** bool_command *)
    let bool_command cmd formula =
      try
        let ic = Unix.open_process_in cmd in
        let formula = Bool.map (fun (_,rex) -> (rex, ref false)) formula in
        (*with_output_to_file "unfiltered.txt" (fun f -> dump_document f doc); *)
        try
          while true do
            let l = input_line ic in
            ignore (Bool.iter (fun (rex, result) ->
              if Pcre.pmatch ~rex l then
                begin
                  result := true
                end) formula)
          done;
          assert false
        with
        | End_of_file ->
            ignore (Unix.close_process_in ic);
            Bool.eval (fun (rex, result) -> !result) formula;
      with
      | x ->
          msg (sf "While executing %S, exception %s" cmd (Printexc.to_string x));
          false
    ;;
    (* ***)
    (*** bool_document *)
    let bool_document url formula =
      if !Opt.debug then
        Printf.eprintf "Checking URL %S\n%!" url;
      let p = new Www.HC.pipeline in
      try
        (*let formula = Bool.map (Pcre.regexp ~study:true) formula in*)
        let doc = Www.obtain_document [] p None url in
        (*with_output_to_file "unfiltered.txt" (fun f -> dump_document f doc); *)
        let res =
          Bool.eval
            (fun (rs,rex) ->
              let result = ref false in
              Www.iter_over_data
                begin fun w ->
                  let outcome = Pcre.pmatch ~rex w in
                  if !Opt.debug then
                    Printf.eprintf "Match(%S,%S)=%b\n%!" rs w outcome;
                  if outcome then
                    begin
                      result := true;
                    end
                end
                doc;
              !result) formula
        in
        p#reset ();
        res
      with
      | x ->
          p#reset ();
          msg (sf "While checking %S, exception %s" url (Printexc.to_string x));
          false
    ;;
    (* ***)
    (*** perform_test *)
    let perform_test e =
       try
         with_timeout e.timeout
           begin
             Bool.eval
               begin function
                 | Url url,formula -> bool_document url formula
                 | Cmd cmd,formula -> bool_command cmd formula
               end
           end
           e.test
       with
       | Alarm -> false
    ;;
    (* ***)
    (*** check *)
    let check tests =
      List.iter
        begin fun e ->
          let t = Unix.gettimeofday () in
          if e.last_test = 0.0 or t -. e.last_test > e.delay then
            begin 
              msg (sf "Testing for %S" e.name);
              e.last_test <- t;
              let result = perform_test e in
              if e.is_good then
                begin
                  if result then
                    begin
                      e.bad_count <- 0;
                      msg (sf "%S is okay" e.name)
                    end
                  else
                    begin
                      e.bad_count <- e.bad_count + 1;
                      if e.bad_count >= e.bad_max then
                        begin
                          msg (sf "Houston, we have a problem on %S" e.name);
                          e.is_good <- false;
                          if not !continue then Action.perform_action e
                        end
                      else
                        msg (sf "There seems to be a problem on %S, count = %d" e.name e.bad_count);
                    end
                end
              else
                begin
                  if result then
                    begin
                      msg (sf "%S is good again" e.name);
                      e.bad_count <- 0;
                      e.is_good <- true;
                      if not !continue then Action.perform_action e
                    end
                  else
                    begin
                      e.bad_count <- e.bad_count + 1;
                      if e.bad_count >= e.bad_max then
                        begin
                          msg (sf "Houston, we still have a problem on %S" e.name);
                          e.is_good <- false;
                        end
                      else
                        msg (sf "There seems to be a problem on %S, count = %d" e.name e.bad_count);
                    end
                end
            end
          else
            ()
        end
        tests
    ;;
    (* ***)
    (*** forever *)
    let forever tests =
      Sys.set_signal
        Sys.sigalrm
        (Sys.Signal_handle(fun _ -> raise Alarm));
      while true do
        check tests;
        continue := false;
        Unix.sleep 5
      done
    ;;
    (* ***)
    (*** dump_document *)
    let dump_document url =
      let p = new Www.HC.pipeline in
      let doc = Www.obtain_document [] p None url in
      (*with_output_to_file "unfiltered.txt" (fun f -> dump_document f doc); *)
      Www.dump_document stdout doc
    ;;
    (* ***)
  (*** testify *)
  let testify 
    (l : (string *
          (Target.source * (string * Pcre.regexp) Bool.boolean) list *
          (int * Target.action list))
            list)
    =
    List.map
      begin fun (name, tests, (bad_max, actions)) ->
        let t = And(List.map (fun x -> Atom x) tests) in
        { default_entry with
          delay = !Opt.delay;
          bad_max = bad_max;
          action = actions;
          name = name;
          test = t }
      end
      l
  ;;
  (* ***)
  end
;;

(*** spec *)
let spec = [
  "-continue", Arg.Set continue, " continue processing ; don't re-send alerts";
  "-dump", Arg.String(fun url -> Do.dump_document url), "<url> Retrieve HTML document and dump its syntax tree.";
  "-config-file", Arg.Set_string Opt.config_file, sf "<path> Set configuration file (%s)" !Opt.config_file;
  "-debug", Arg.Set Opt.debug, " Enable debugging";
  "-grep",
    Arg.Tuple(
      let url = ref "" in
      [Arg.Set_string url;
       Arg.String(fun regex -> Do.grep_document !url regex)]),
    "<url> <regex>  Download and grep document.";
  "-delay", Arg.Set_float Opt.delay, "<delay> Delay between tests in seconds.";
  "-sendmail", Arg.Set_string Opt.sendmail, sf "<sendmail> Set sendmail command (%S)" !Opt.sendmail
]
(* ***)

let _ =
  Arg.parse spec (fun _ -> ()) (sf "Usage: %s [options]" (Filename.basename Sys.argv.(0)));
  let cfg = Parser.load !Opt.config_file in
  let tests = Do.testify cfg in
  Do.forever tests
;;
